### ISSUE TITLE
"http: TLS handshake error from XXXXXXXX:PPPP: tls: client didn't provide a certificat"

### ISSUE THREAD

**Issue by 16bc** [go-gost/gost#579]
I created a certificate for the server domain with 3 pem-files using sertbot -sertonly.
If i pass it as `?certFile=cert.pem&keyFile=key.pem&caFile=ca.pem`
i have error:
`http: TLS handshake error from XXXXXXXX:PPPP: tls: client didn't provide a certificat`
I created client.key and client.crt on client didnt solve that.
I don't want the GOST to use the autocreated certificate with the name gost.run
Help!

**Comment by ginuerzh** [comment]
Hi @16bc,

There are two separate things going on here — neither is a bug, but the certificate handling is subtle. Let me break it down.

**1. `caFile` on the server turns on mutual TLS (mTLS).**

When you pass `caFile` to a listener (server side), gost loads it as the *client* CA and requires every client to present a certificate signed by it (`ClientAuth = RequireAndVerifyClientCert`). So this:

```
gost -L "tls+mwss://:443?certFile=cert.pem&keyFile=key.pem&caFile=ca.pem"
```

tells the server to require a client certificate signed by `ca.pem`. A browser or any client without one is rejected with exactly your error: `tls: client didn't provide a certificate`. The certbot `ca.pem` is Let's Encrypt's chain — meant for a *client* to verify your server, not to sign client certs.

If you only want an **encrypted tunnel** (one-way TLS), drop `caFile` from the server and keep just `certFile` + `keyFile`:

```bash
# Server — one-way TLS, no client cert required
gost -L "mwss://:443?certFile=fullchain.pem&keyFile=privkey.pem"
# Client — verify the server cert (Let's Encrypt is in the system root store)
gost -L socks5://:1030 -F "mwss://your-domain:443?secure=true"
```

Use `fullchain.pem` (cert + intermediate), not the bare leaf, so clients can build the chain.

**2. `tls+mws` is plaintext, not "TLS over mws".**

The `+` separates *handler* and *listener* (`handler+listener://`); it is not a TLS wrapper. `tls+mws` is parsed as handler `tls` (not a real handler, so it falls back to `auto`) over the plaintext `mws` listener. The result is **no TLS at all** — your `certFile`/`keyFile`/`caFile` are silently ignored. That's why the client "works without keys" (no handshake to enforce) and a browser over HTTPS gets `SSL_ERROR_RX_RECORD_TOO_LONG`.

The TLS-enabled variant is `mwss://` (trailing `s` = secure); likewise `ws`→`wss`. Use `mwss://` directly — no `tls+` prefix needed.

**TL;DR**
- Encrypted tunnel: use `mwss://` with only `certFile` + `keyFile` on the server; add `secure=true` on the client to verify the cert.
- Only add `caFile` on the server if you want mTLS (every client must present a cert signed by a CA you control — not Let's Encrypt's).
- `tls+<scheme>` does not add TLS; use the `...s` listener variant.

Side note: `probeResistance` is a feature of the HTTP proxy handler (`http`/`http2`), used to disguise the proxy from unauthenticated probes. It doesn't apply to a plain `mwss` + `socks5` setup.

(Heads-up: in the latest `x`, a warning is now logged when TLS options are configured for a listener that ignores them, so this misconfiguration is no longer silent.)

Hope this helps!
